File: Packages/WhoopProtocol/Sources/WhoopProtocol/StandardHRReceipt.swift
import Foundation

/// Original 0x2A37 notification and host receipt clocks. These are arrival clocks, NOT beat timestamps.
/// A session/ordinal identifies a notification even when several identical packets arrive in one second.
/// Neither session membership nor consecutive ordinals establishes uninterrupted sensor acquisition.
public struct StandardHRReceipt: Codable, Equatable, Sendable {
    public let receiptId: String
    public let ts: Int
    public let sessionId: String
    public let notificationOrdinal: Int64
    public let receivedUnixMs: Int64
    public let receivedMonotonicNs: Int64
    public let rawHex: String
    public let schemaVersion: Int
    public let clockVersion: String

    public static func capture(_ bytes: [UInt8], sessionId: String, notificationOrdinal: Int64,
                               receivedUnixMs: Int64, receivedMonotonicNs: Int64) -> Self? {
        guard UUID(uuidString: sessionId) != nil, notificationOrdinal >= 0,
              receivedUnixMs >= 0, receivedMonotonicNs >= 0, (1...512).contains(bytes.count) else { return nil }
        let id = sessionId.lowercased()
        return Self(receiptId: "\(id):\(notificationOrdinal)", ts: Int(receivedUnixMs / 1000),
            sessionId: id, notificationOrdinal: notificationOrdinal, receivedUnixMs: receivedUnixMs,
            receivedMonotonicNs: receivedMonotonicNs,
            rawHex: bytes.map { String(format: "%02x", $0) }.joined(), schemaVersion: 1,
            clockVersion: "host-arrival-unmapped")
    }

    public var isValid: Bool {
        guard let bytes = RRPacketProvenance.bytes(rawHex) else { return false }
        return Self.capture(bytes, sessionId: sessionId, notificationOrdinal: notificationOrdinal,
            receivedUnixMs: receivedUnixMs, receivedMonotonicNs: receivedMonotonicNs) == self
    }

    /// Exact, ordered interval words including zero/invalid physiological values. A truncated declared
    /// field invalidates the decoded array, while the original bytes remain retained for diagnosis.
    public var rrRawTicks: [Int]? {
        guard isValid, let bytes = RRPacketProvenance.bytes(rawHex), let flags = bytes.first else { return nil }
        var offset = flags & 1 == 0 ? 2 : 3
        guard bytes.count >= offset else { return nil }
        if flags & 8 != 0 { offset += 2 }
        guard bytes.count >= offset else { return nil }
        guard flags & 16 != 0 else { return bytes.count == offset ? [] : nil }
        guard bytes.count > offset, (bytes.count - offset) % 2 == 0 else { return nil }
        return stride(from: offset, to: bytes.count, by: 2).map { Int(bytes[$0]) | Int(bytes[$0 + 1]) << 8 }
    }
}

File: android/app/src/main/java/com/noop/protocol/StandardHrReceipt.kt
package com.noop.protocol

import java.util.UUID

/** Original 0x2A37 bytes plus host ARRIVAL clocks, never a verified beat timestamp or continuity claim. */
data class StandardHrReceipt(
    val receiptId: String, val ts: Long, val sessionId: String, val notificationOrdinal: Long,
    val receivedUnixMs: Long, val receivedMonotonicNs: Long, val rawHex: String,
    val schemaVersion: Int = 1, val clockVersion: String = "host-arrival-unmapped",
) {
    val isValid: Boolean get() = RrPacketProvenance.bytes(rawHex)?.let {
        capture(it, sessionId, notificationOrdinal, receivedUnixMs, receivedMonotonicNs) == this
    } == true

    /** Preserve zeros/invalid physiological values and original slot order; reject incomplete fields. */
    val rrRawTicks: List<Int>? get() {
        if (!isValid) return null
        val bytes = RrPacketProvenance.bytes(rawHex) ?: return null
        val flags = bytes[0].toInt() and 255
        var offset = if (flags and 1 == 0) 2 else 3
        if (bytes.size < offset) return null
        if (flags and 8 != 0) offset += 2
        if (bytes.size < offset) return null
        if (flags and 16 == 0) return if (bytes.size == offset) emptyList() else null
        if (bytes.size <= offset || (bytes.size - offset) % 2 != 0) return null
        return (offset until bytes.size step 2).map { (bytes[it].toInt() and 255) or ((bytes[it + 1].toInt() and 255) shl 8) }
    }

    companion object {
        fun capture(bytes: ByteArray, sessionId: String, notificationOrdinal: Long,
                    receivedUnixMs: Long, receivedMonotonicNs: Long): StandardHrReceipt? {
            val id = runCatching { UUID.fromString(sessionId).toString() }.getOrNull() ?: return null
            if (!id.equals(sessionId, ignoreCase = true) || notificationOrdinal < 0 || receivedUnixMs < 0 ||
                receivedMonotonicNs < 0 || bytes.size !in 1..512) return null
            val hex = bytes.joinToString("") { "%02x".format(it.toInt() and 255) }
            return StandardHrReceipt("$id:$notificationOrdinal", receivedUnixMs / 1000, id,
                notificationOrdinal, receivedUnixMs, receivedMonotonicNs, hex)
        }
    }
}
